### PR TITLE
Added an example to deploy all Edge3.0 components using EIB

### PR DESCRIPTION
**Pull request by e-minguez**
This PR includes all the Edge 3.0 components deployed via EIB:

* RKE2 v1.28.8 with Cilium and multus as CNI as well as enabling SELinux (on 3 nodes)
* Rancher Prime 2.8.3 (as well as performing the bootstrap automatically via API calls to Rancher)
* metal3 (using metallb to expose the Ironic endpoint as a LoadBalancer) + CAPI (via clusterctl)
* Longhorn 1.6.0
* NeuVector 5.3.2 (with some tweaks to fit Edge scenarios)
* Cert-manager 1.14.2 (as requisite for both metal3 and Rancher)
* Kubevirt (& CDI)
* Elemental
* Rancher UI Plugin operator
* Elemental Rancher Plugin
* Kubevirt Rancher Plugin

Those are deployed via EIB helm charts feature (aka RKE2 Addons) and for metal3 and Rancher, there are scripts that perform the required customization directly on the host.

There are as some quality of life improvements as a bonus:
* Hosts are registered to SCC (via a custom script in combustion)
* `qemu-guest-agent` is installed (and enabled) for VM communication with the host
* Disabled automatic updates & reboots (`rebootmgr.service`,  `transactional-update.timer` and `transactional-update-cleanup.timer`) to prevent unexpected situations.
* Added some handy aliases (such as `alias k=kubectl`) to the shell
* rke2-ingress-nginx is configured to be exposed via metallb (via a `HelmChartConfig`)
* A `hello-node` example application is deployed as well

### CODE REVIEW

**Comment by e-minguez** [comment]
I'll add a README with instructions, vars, etc. once we agree everything is in place.

**Review by fdegir — COMMENTED** [review]
Few comments inline.

File: examples/edge30-full-example/eib.yaml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    initializer: true
+    type: server
+  - hostname: host2rke2.example.com
+    type: server
+  - hostname: host3rke2.example.com
+    type: server
+  version: v1.28.8+rke2r1
+operatingSystem:
+  isoConfiguration:
+    installDevice: /dev/sda
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge30-full-example/eib.yaml:125]: Will do.

File: examples/edge30-full-example/eib.yaml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    - name: jetstack
+      url: https://charts.jetstack.io
+    - name: rancher-charts
+      url: https://charts.rancher.io/
+    - name: suse-edge-charts
+      url: https://suse-edge.github.io/charts
+    - name: elemental-operator-crd
+      url: oci://registry.suse.com/rancher/elemental-operator-crds-chart
+    - name: elemental-operator
+      url: oci://registry.suse.com/rancher/elemental-operator-chart
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge30-full-example/eib.yaml:102]: There's been too many changes in the last few days... I need to update this one as well, thanks.

File: examples/edge30-full-example/eib.yaml [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      version: 1.3.0
+    repositories:
+    - name: jetstack
+      url: https://charts.jetstack.io
+    - name: rancher-charts
+      url: https://charts.rancher.io/
+    - name: suse-edge-charts
+      url: https://suse-edge.github.io/charts
+    - name: elemental-operator-crd
+      url: oci://registry.suse.com/rancher/elemental-operator-crds-chart
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge30-full-example/eib.yaml:100]: There's been too many changes in the last few days... I need to update this one as well, thanks.

File: examples/edge30-full-example/eib.yaml [hunk 4]
```diff
@@ -0,0 +1,143 @@
+apiVersion: 1.0
+image:
+  arch: x86_64
+  baseImage: SLE-Micro.x86_64-5.5.0-Default-SelfInstall.install.iso
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge30-full-example/eib.yaml:4]: Correct.

**Comment by e-minguez** [comment]
All comments addressed @fdegir PTAL. Thanks.

File: examples/edge30-full-example/custom/files/rancher.sh [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  # If there is no final password, then finish the setup right away
+  exit 0
+fi
+
+trap 'catch $? $LINENO' EXIT
+
+catch() {
+  if [ "$1" != "0" ]; then
+    echo "Error $1 occurred on $2"
+    ${KUBECTL} delete configmap ${METAL3LOCKCMNAME} -n ${METAL3LOCKNAMESPACE}
```
[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge30-full-example/custom/files/rancher.sh:20]: correct, my bad

[the remark this answers is not in the document]
Comment (e-minguez) [reply examples/edge30-full-example/custom/files/rancher.sh]: Fixed, PTAL, thanks.

File: examples/edge30-full-example/custom/files/basic-setup.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+export METAL3_CLUSTERCTLVERSION="1.6.2"
+export METAL3_CAPICOREVERSION="1.6.0"
+export METAL3_CAPIMETAL3VERSION="1.6.0"
+export METAL3_CAPIRKE2VERSION="0.2.6"
+export METAL3_CAPIPROVIDER="rke2"
+export METAL3_CAPISYSTEMNAMESPACE="capi-system"
+export METAL3_RKE2BOOTSTRAPNAMESPACE="rke2-bootstrap-system"
+export METAL3_CAPM3NAMESPACE="capm3-system"
+export METAL3_RKE2CONTROLPLANENAMESPACE="rke2-control-plane-system"
+export METAL3_CAPI_IMAGES="registry.opensuse.org/isv/suse/edge/clusterapi/containerfile/suse"
```
Comment (alknopfler) [inline examples/edge30-full-example/custom/files/basic-setup.sh:19]: @e-minguez requested downstream change

Comment (e-minguez) [reply examples/edge30-full-example/custom/files/basic-setup.sh]: Fixed, thanks!

File: examples/edge30-full-example/custom/files/metal3.sh [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  # Try this command 3 times just in case, stolen from https://stackoverflow.com/a/33354419
+  if ! (r=3; while ! clusterctl init \
+    --core "cluster-api:v${METAL3_CAPICOREVERSION}"\
+    --infrastructure "metal3:v${METAL3_CAPIMETAL3VERSION}"\
+    --bootstrap "${METAL3_CAPIPROVIDER}:v${METAL3_CAPIRKE2VERSION}"\
+    --control-plane "${METAL3_CAPIPROVIDER}:v${METAL3_CAPIRKE2VERSION}" ; do
+            ((--r))||exit
+            echo "Something went wrong, let's wait 10 seconds and retry"
+            sleep 10;done) ; then
+      echo "clsuterctl failed"
```
Comment (alknopfler) [inline examples/edge30-full-example/custom/files/metal3.sh:125]:
```suggestion
      echo "clusterctl failed"
```
```suggestion
      echo "clusterctl failed"
```

File: examples/edge30-full-example/custom/files/basic-setup.sh [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+# Pre-requisites. Cluster already running
+export KUBECTL="/var/lib/rancher/rke2/bin/kubectl"
+export KUBECONFIG="/etc/rancher/rke2/rke2.yaml"
+
+##################
+# METAL3 DETAILS #
+##################
+export METAL3_CHART_TARGETNAMESPACE="metal3-system"
+export METAL3_CLUSTERCTLVERSION="1.6.2"
+export METAL3_CAPICOREVERSION="1.6.0"
```
Comment (alknopfler) [inline examples/edge30-full-example/custom/files/basic-setup.sh:11]: @e-minguez should be 1.6.2 ??? -> "cluster-api:v1.6.2" --infrastructure "metal3:v1.6.0"

**Review by alknopfler — CHANGES_REQUESTED** [review]

File: examples/edge30-full-example/custom/files/basic-setup.sh (hunk 8 above)
Comment (e-minguez) [reply examples/edge30-full-example/custom/files/basic-setup.sh]: Done, thanks.

**Review by alknopfler — APPROVED** [review]
LGTM

**Review by fdegir — APPROVED** [review]
LGTM

Two minor comments which can be addressed separately are

- directory structure under examples folder could perhaps be adjusted, clearly highlighting basic and advanced examples with an update to readme under examples folder.
- a readme file for advanced example.

**Comment by e-minguez** [comment]
Summoning @atanasdinov and @jdob for the folder structure suggestion. I'll add a README for this if we all agree on merging this (honestly I don't want this to happen https://github.com/suse-edge/edge-image-builder/pull/204 again :D )

**Review by atanasdinov — COMMENTED** [review]
@e-minguez This is an impressive contribution. I'd only suggest to split up the directory structure as already suggested by @fdegir. Nobody is going to deploy all Edge 3.0 pieces together at the same time. It's much cleaner and easier to follow if we have the individual projects split into their own "little" examples. Something like:
```
examples/
   basic/
      <the ones we currently have>
   advanced/
       rancher/
           ...
       metal3/
           ...
```

**Comment by e-minguez** [comment]
> @e-minguez This is an impressive contribution. I'd only suggest to split up the directory structure as already suggested by @fdegir. Nobody is going to deploy all Edge 3.0 pieces together at the same time. 

I do :D

> It's much cleaner and easier to follow if we have the individual projects split into their own "little" examples. Something like:
> 
> ```
> examples/
>    basic/
>       <the ones we currently have>
>    advanced/
>        rancher/
>            ...
>        metal3/
>            ...
> ```

Do we prefer to use different 'scenarios' instead? I'm asking because there are dependencies that need to be resolved. For example metal3 requires cert-manager. Rancher requires cert-manager as well. Elemental requires Rancher (hence, cert-manager). If we want metal3+rancher, we need to disable some stuff https://github.com/suse-edge/edge-image-builder/pull/378/files#diff-f3860ea69cf5ae7c40c0a2b4d0a8c5ea64a9fba4cb13edf3f17d76ae853e1472R88

In any case, I still think having a _full_ example will be beneficial... at the end of the day all the others will be trimmed versions of this.

**Comment by atanasdinov** [comment]
I don't know what to call those scenarios, though. It's precisely the case that the main components have some dependencies that if a user wants to deploy those, they'd be able to very easily see what's necessary.

On a separate note, why is Rancher a hard requirement for Elemental? We've deployed it without it.

**Comment by e-minguez** [comment]
> I don't know what to call those scenarios, though. It's precisely the case that the main components have some dependencies that if a user wants to deploy those, they'd be able to very easily see what's necessary.

In the elemental (+ui) example, shall we call ui and rancher optional or have two different scenarios? Rancher + Elemental + Elemental UI and another one with just Elemental?

> On a separate note, why is Rancher a hard requirement for Elemental? We've deployed it without it.

It is not sorry, I was throwing the elemental ui on the same bucket :)

**Comment by e-minguez** [comment]
I'm closing this due to inactivity. Perhaps having smaller examples is better.

**Pull request closed without merging**